### PR TITLE
Generate DSC Metaconfig

### PR DESCRIPTION
**Pull request by hestolz**
CRI 177130188
ADO 6369194

Need to regenerate DSC metaconfig when we get DSC endpoint in topology request response. Testing methodology:
- emplace modified `agent_maintenance_script.rb`
- `/opt/microsoft/omsagent/bin/service_control restart`
- modify `DSC_ENDPOINT` in `omsadmin.conf` with fake value
- `su - omsagent -c /opt/microsoft/omsconfig/Scripts/OMS_MetaConfigHelper.py`
- check that `/etc/opt/omi/conf/omsconfig/generated_meta_config.mof` has new fake value
- wait for topology request, check `generated_meta_config.mof` again for replaced, corrected value

### CODE REVIEW

**Comment by Bhargava-Chary-Chollaty** [comment]
Can we consider adding log when metaconfig.mof is re-generated and what was the change in endpoint. Old and Updated endpoint.

File: source/code/plugins/agent_maintenance_script.rb [hunk 1]
```diff
@@ -190,9 +191,44 @@ def apply_dsc_endpoint(server_resp)
...
 
+      generate_dsc_metaconfig
+
       return dsc_endpoint
     end
 
+    # Generate the DSC MetaConfig file. If the dsc endpoint is updated, DSC will
+    # not directly pick the endpoint from omsadmin.conf (instead, it is picked
+    # from MetaConfig.mof), so we must manually trigger metaconfig (re)generation.
+    def generate_dsc_metaconfig
```
Comment (rishabh1jain) [inline source/code/plugins/agent_maintenance_script.rb]: Umm will it more efficient to check if it has changed, rather than killing and restarting the process each time a topology response is received i.e. every hour?

Comment (hestolz) [reply source/code/plugins/agent_maintenance_script.rb]: the process will not be killed

Comment (rishabh1jain) [reply source/code/plugins/agent_maintenance_script.rb]: Oh I see, you are running the command and monitoring that. my bad.

Comment (hestolz) [reply source/code/plugins/agent_maintenance_script.rb]: No worries—basically we are running a .py script that will write to `generated_meta_config.mof`. Neither DSC nor OMS needs to be restarted for the change to that file to take effect. Since we run this as a subprocess, we need to make sure it doesn't  hang or turn zombie. For instance, if someone else is has already taken the DSC lock, this call could fail or take some time. In that case we shouldn't wait forever. Next time topology request is made we will try again.

**Comment by hestolz** [comment]
> Can we consider adding log when metaconfig.mof is re-generated and what was the change in endpoint. Old and Updated endpoint.

I added a log statement if any value changes in `omsadmin.conf`. In all cases `omsconfig.log` will show the result when we do these metaconfig generation calls.

**Review by chsamala — COMMENTED** [review]
- Because  of locking if you failed to apply new metaconfiguration , then please retry rather than waiting for next topology request.

- if there is no change in topology configuration(DSC part) then dont apply the meta configuration.

File: source/code/plugins/agent_maintenance_script.rb [hunk 2]
```diff
@@ -190,13 +196,49 @@ def apply_dsc_endpoint(server_resp)
...
+            begin
+              `pkill -TERM -g #{pgid}`
+              sleep 1
+              if wait_thr.status
+                `pkill -KILL -g #{pgid}`
+              end
+            rescue => e
+              log_error("Failed to kill process(es) for command '#{command}'. #{e}")
+            end
+            OMS::Log.warn_once("DSC MetaConfig generation failed due to subprocess timeout")
```
[the remark this answers is not in the document]
Comment (hestolz) [reply source/code/plugins/agent_maintenance_script.rb:227]: I have tested this and it works. I already detailed how I tested this in the PR description.

If you check the OMS_MetaConfigHelper.py it doesn't need any parameters if it is being run to generate metaconfig. The parameters are only necessary if you are trying to disable or enable DSC config management.

[the remark this answers is not in the document]
Comment (Bhargava-Chary-Chollaty) [reply source/code/plugins/agent_maintenance_script.rb:227]: OMS_MetaConfigHelper.py persists the argument/flag that it was executed with previously by creating file (/etc/opt/omi/conf/omsconfig/omshelper_disable).

Assume a user runs OMS_MetaConfigHelper.py with --disable flag. This would change the Metaconfig from PULL mode to PUSH mode. If the user runs the script second time with out any flag, then the script will still use previous argument and configures Metaconfig in PUSH mode where DSC endpoint is not configured.

We need to test the changes considering scenario where the DSC endpoint change is happened when the omsagent is in disables state and is enabled later by the user.

File: source/code/plugins/agent_maintenance_script.rb (hunk 1 above)
Comment (Bhargava-Chary-Chollaty) [reply source/code/plugins/agent_maintenance_script.rb:207]: I think the mof file to be checked if the change is successfully applied is at `/etc/opt/omi/conf/omsconfig/configuration/MetaConfig.mof`

`generated_meta_config.mof` is an intermediate file generated which is sent to dsc while applying meta config.

File: source/code/plugins/agent_maintenance_script.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (hestolz) [reply source/code/plugins/agent_maintenance_script.rb]: I've added the check to see if DSC is disabled. If so, I won't call `OMS_MetaConfigHelper.py` but will persist a flag so once DSC is re-enabled we will call.

**Comment by hestolz** [comment]
Addressed Feedback
- Retry up to 3 times
- Only generate the metaconfig when there is a change
- Add logging indicating when there is a change, and if so log the old/new values
- Handle possible DSC disabled case

File: source/code/plugins/agent_maintenance_script.rb [hunk 3]
```diff
@@ -171,32 +183,105 @@ def apply_certificate_update_endpoint(server_resp, check_for_renew_request = tru
...
+        end
+      end
+
+      OMS::Log.warn_once("Exceeded max attempts to generate DSC metaconfig, will retry after next topology request")
+    end
+
+    # Call OMS_MetaConfigHelper.py to trigger metaconfig generation, return success
+    def call_oms_metaconfighelper
+      # Note that OMS_MetaConfigHelper.py will log to omsconfig.log
+      command = "/opt/microsoft/omsconfig/Scripts/OMS_MetaConfigHelper.py"
```
Comment (hestolz) [inline source/code/plugins/agent_maintenance_script.rb:246]: this will have to change once [this](https://github.com/microsoft/PowerShell-DSC-for-Linux/pull/691/files) is taken

File: source/code/plugins/agent_maintenance_script.rb [hunk 4]
```diff
@@ -171,32 +188,105 @@ def apply_certificate_update_endpoint(server_resp, check_for_renew_request = tru
...
+        if success
+          @pending_metaconfig_generation = false
+          return
+        else
+          sleep 5
+          retries += 1
+        end
+      end
+
+      log_warn("Exceeded max attempts to generate DSC metaconfig, will retry after next topology request")
```
Comment (Bhargava-Chary-Chollaty) [inline source/code/plugins/agent_maintenance_script.rb:245]: If all the retries are exhausted, we are saying that the dsc endpoint is updated in next topology request. 

In the first request, we call the update_config with new endpoint which i assume will update the omsadmin.conf. If all the retries fail we never revert this change, how will we able to identify the change in endpoint in this scenario for next topology requests ?

File: source/code/plugins/oms_configuration.rb [hunk 5]
```diff
@@ -365,6 +365,15 @@ def set_request_intervals(topology_interval, telemetry_interval)
         OMS::Log.info_once("OMS agent management service telemetry request interval now #{@@TelemetryInterval}")
       end
 
+      # Check whether DSC has been disabled (which prevents automatic config management)
+      def is_dsc_disabled
+        begin
+          return File.exist?("/etc/opt/omi/conf/omsconfig/omshelper_disable")
+        rescue => e
+          OMS::Log.error_once("Failed to get DSC enabled/disabled status. #{e}")
```
Comment (Bhargava-Chary-Chollaty) [inline source/code/plugins/oms_configuration.rb:373]: We do not return a boolean to caller on exception. How would this effect the flow ?

File: source/code/plugins/agent_maintenance_script.rb (hunk 4 above)
Comment (hestolz) [reply source/code/plugins/agent_maintenance_script.rb:245]: Think of it in terms of invariants

1. omsadmin.conf will always be updated with the newest DSC_ENDPOINT following every topology request
2. pending_metaconfig_generation will always be set when a change is detected *and never unset until the metaconfig is successfully generated*.

Since the flag is never unset there is no reason to revert the change to omsadmin.conf; when the next topology request comes around metaconfig generation will be retried.

Comment (Bhargava-Chary-Chollaty) [reply source/code/plugins/agent_maintenance_script.rb:245]: Would it mean the state of variable 'pending_metaconfig_generation' is always persisted. Even for different executions of the script. I am assuming that this is not executed by a continuous process and there is an initialization of the variable to false in the start. Sorry if it is obvious, I did not get how it works in rb.

Comment (hestolz) [reply source/code/plugins/agent_maintenance_script.rb:245]: No problem. It is persisted because this file defines a class `Maintenance` which is instantiated in `in_heartbeat_request.rb`, which is an input plugin that starts up and runs for the entire life cycle of the agent. You can see in that file we call `@maintenance_script.heartbeat` (do a topology request) and `@maintenance_script.generate_dsc_metaconfig` (call this method).

Comment (Bhargava-Chary-Chollaty) [reply source/code/plugins/agent_maintenance_script.rb:245]: Wouldn't machine restart effect this ?

**Comment by hestolz** [comment]
Thank you for your submission. As [previously announced](https://azure.microsoft.com/en-us/updates/were-retiring-the-log-analytics-agent-in-azure-monitor-on-31-august-2024/), the Log Analytics agent has been **deprecatated** and has no support as of **August 31, 2024.** If you use the Log Analytics agent to ingest data to Azure Monitor, [migrate now to the new Azure Monitor agent](https://docs.microsoft.com/en-us/azure/azure-monitor/agents/azure-monitor-agent-migration). As part of repo archival, open issues and pull requests will be closed.
